canard.cpp

```diff
@@ -0,0 +1,186 @@
+#include <libcanard/canard.h>
+#include <cstdint>
+#include <libcanard/drivers/stm32/canard_stm32.h>
+#include <libcanard/helpers/RingBuffer.h>
+#include <libcanard/helpers/helpers.h>
+
+static RingBuffer<CanardCANFrame, CAN_RX_RB_SIZE> CAN_RX_RB;
+static RingBuffer<CanardCANFrame, CAN_TX_RB_SIZE> CAN_TX_RB;
+
+int getRXSize() {
+    return CAN_RX_RB.size();
+}
+
+int getTXSize() {
+    return CAN_TX_RB.size();
+}
+
+
+extern "C" {
+extern CanardInstance canard;
+
+volatile int allTimeMaxRxSize;
+volatile int allTimeMaxTxSize;
+volatile int maxRxSize;
+volatile int maxTxSize;
+
+#if defined(STM32F0) || defined(STM32F3)
+void USED CAN_IT_Callback() {
+    if (!CAN_RX_RB.is_full()) {
+        canardSTM32Receive(&CAN_RX_RB.get_next_write_entry());
+        CAN_RX_RB.advance_write();
+    } else {
+        canardSTM32ReleaseFIFO();
+        canard_errors.rx_errors++;
+    }
+
+    volatile int rxSize = getRXSize();
+    if (rxSize > maxRxSize) {
+        maxRxSize = rxSize;
+    }
+    if (rxSize > allTimeMaxRxSize) {
+        allTimeMaxRxSize = rxSize;
+    }
+    canardSTM32Receive_IT();
+}
+#endif
+
+volatile bool canTXTransmitting = false;
+
+volatile canard_errors_t canard_errors = {
+    .tx_errors = 0,
+    .rx_errors = 0,
+    .broadcast_errors = 0
+};
+
+void processRxQueue() {
+    const uint64_t ts_usec = getMonotonicTimestampMSec() * 1000UL;
+
+    while (!CAN_RX_RB.is_empty()) {
+        canardHandleRxFrame(&canard, &CAN_RX_RB.get_next_read_entry(), ts_usec);
+        CAN_RX_RB.advance_read();
+    }
+}
+
+void processTxQueue(void) {
+    /* Call transmit function */
+    if (!CAN_TX_RB.is_empty()) {
+        CanardCANFrame &frame = CAN_TX_RB.get_next_read_entry();
+
+        canTXTransmitting = true;
+        const int tx_res = canardSTM32Transmit(&frame);
+        if (tx_res < 0) {         // Failure - drop the frame and report
+            canard_errors.tx_errors++;
+            canTXTransmitting = false;
+        } else if (tx_res == 0) {
+            canard_errors.tx_errors++;
+            canTXTransmitting = false;
+        } else {
+            CAN_TX_RB.advance_read();
+        }
+
+        volatile int txSize = getTXSize();
+        if (txSize > maxTxSize) {
+            maxTxSize = txSize;
+        }
+        if (txSize > allTimeMaxTxSize) {
+            allTimeMaxTxSize = txSize;
+        }
+
+    } else {
+        canTXTransmitting = false;
+    }
+}
+
+void processTxRxOnce() {
+    NVIC_DisableIRQ(CANARD_TX_IRQn);
+    __ISB();
```

Comment:
These memory barriers ought to be rolled into the canardSTM32{En,Dis}ablePeripheral functions for readability.

Reply:
Or removed, it's not entirely clear that they're necessary

Reply:
I'm assuming it's because of the "Implementation requirements" section of http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dai0321a/BIHHFHJD.html

Looks like newer CMSIS versions [roll these barriers into `NVIC_DisableIRQ`](https://github.com/ARM-software/CMSIS_5/commit/d1a283ace989eadf7e3298d1219de62fe40270d4)